### DISCUSSION TITLE
Wizard flicker / starts over

### DISCUSSION THREAD

**Discussion by rabol** [spatie/laravel-livewire-wizard#77] [Q&A]
Hi

I have noticed that while developing / manual testing a wizard it will flicker or start over if you modify the blade template

let's say that you have a 3-4 step wizard and you do modification to the step 3-4, then when you reach one of these steps the wizard will start over.

is there a way completely reset the state ?

**Comment by ju5t** [comment]
Every time? Or are you using something like Browsersync?

**Reply by rabol** [reply]
yes everytime I have made a modification to the blade file - I suspect that it could be Vite?

**Reply by ju5t** [reply]
Could be, if you're combining it with Browsersync to refresh the browser automatically. I don't know. There wouldn't be a way I know of to reset the state when something like this happens though.

**Reply by rabol** [reply]
I do not have browsersync

**Reply by ju5t** [reply]
Then it's probably either through a plugin in Vite (`handleHotUpdate`) or you've enabled it in the [Vite config](https://laravel.com/docs/10.x/vite#blade-refreshing-on-save). I don't know how it works in Vite, but there are often two reloading mechanisms.

1. It will use some sort of dom diffing, like Livewire does, but not the same. This will refresh the page without a browser refresh.
2. Or it will just refresh your browser. As if you would press F5.

If it's option 1, there isn't much you can do about it. I don't know of a way to hook into the refresh and reset the state. I would disable the refresh mechanism while you're debugging the wizard and refresh manually. 

If it's option 2, it should work just fine. There shouldn't be a flicker and it would always go back to the wizard's initial state.

**Reply by topegret** [reply]
I have the same issue.

**Reply by alignwebs** [reply]
I have the same issue too

**Reply by ju5t** [reply]
Without more information it's impossible to give any help.

**Reply by rabol** [reply]
The issue is related to Vite - IMHO

I get the flickering when: npm run dev, and tghen the css/js is hot loaded by the Vite 'server'

when building the css/js and not have the Vite server running it's all okay

not sure who to blame, but AFIK, normal Livewire components does not have this issue

**Reply by ju5t** [reply]
I can't replicate it on my end. Can you share your vite.config.js? Which versions are you using?

Edit:
In the meantime I have done a few tests.

An app of ours was on Vite 2.9 and an old Laravel plugin and that didn't cause any issues. I then updated all versions to the latest ones and the only thing that happens is that it appears to do a hard refresh and moves it back to the first step (undoing all changes I made along the way). This is what I would expect from such a refresh. It doesn't flicker here.

This is with Vite 4 and plugin version 0.7.

```javascript
export default defineConfig({
    plugins: [
        laravel({
            input: [
                // Your files
            ],
            refresh: true,
        }),
    ],
})
```

**Reply by rabol** [reply]
What you are describing is what I call 'Flicker', however I goes back to first step several times, even if you are not in the wizard .

e.g: modify the blade for one of the steps, then see Vite do a 'reload', then start the wizard and after going to 'next' it goes back to the start.

Now, if I was 'in the' wizard when the vite reload happens, I somehow could understand.



```
import { defineConfig } from 'vite';
import laravel from 'laravel-vite-plugin';
import {resolve} from "path";
import {homedir} from "os";
import fs from "fs";
import { viteStaticCopy } from 'vite-plugin-static-copy'

let host = 'sign.test';
export default defineConfig({
    plugins: [
        laravel({
            input: [
                'resources/css/app.css',
                'resources/js/app.js',

            ],
            refresh: true,

        }),
    ],

    build: {
        minify: true,
    },

    server: detectServerConfig(host),
});

function detectServerConfig(host) {
    let keyPath = resolve(homedir(), `.config/valet/Certificates/${host}.key`)
    let certificatePath = resolve(homedir(), `.config/valet/Certificates/${host}.crt`)

    if (!fs.existsSync(keyPath)) {
        return {}
    }

    if (!fs.existsSync(certificatePath)) {
        return {}
    }

    return {
        hmr: {host},
        host,
        https: {
            key: fs.readFileSync(keyPath),
            cert: fs.readFileSync(certificatePath),
        },
    }
}
```
php version 8.1.17
laravel version 10.6.1
composer version: 2.5.3
Vite version 4.2.1, plugin v0.7.4
spatie/laravel-livewire-wizard v1.1.5

**Reply by ju5t** [reply]
To elaborate why it is normal for the wizard to reset and go back to step one: the wizard stores data in a Livewire variable. This means it's volatile until you either store it in a session or the database. It's the same as most Livewire components, if you do not store changes automatically, a refresh will run `mount()` and refresh its state.

**Reply by ju5t** [reply]
No matter what I try I cannot replicate the scenario where:

1. You make a change.
2. Go to the wizard.
3. Click on next and it goes back to the first step.

But, if I am inside a wizard and make a change it will do a hard refresh. The 'flickering' you may see it the browser reloading the entire page. This seems to be default Vite, the alternative is HMR. That's what `handleHotUpdate` should do according to the docs in Vite, but it doesn't do that when I switch back to it.

Also, Vite kind of gives it away in their prompt:

```bash
[vite] page reload
```

Conclusion: the hard refresh is what you would expect from the configuration you have and that would always do a full reload. The odd behaviour you're seeing isn't normal though but I do not think it's related to the wizard. You can try to update to 1.2.x, which is what we're on, but I don't think it will solve it.

Edit: I see you've enabled HMR in you're host settings, but I'm unsure if Blade is supported. I think it will only hot swap resources that are defined like `@vite('resources/css/app.css')`. Everything else will cause a page reload.

**Comment by waapatricio** [comment]
I'm going through the same problem. When modifying something in the components, during the tests the vite is updating and returning to the beginning.

`1:44:45 PM [vite] page reload app/Http/Livewire/Pages/CheckoutPage.php (x5)
1:44:45 PM [vite] page reload app/Http/Livewire/Components/Checkout/CheckoutWizardComponent.php
1:44:52 PM [vite] page reload storage/framework/views/0cfbc7d9c1fae675feecf591553f4e67.php
1:44:52 PM [vite] page reload storage/framework/views/0cfbc7d9c1fae675feecf591553f4e67.php (x2)
1:44:53 PM [vite] page reload storage/framework/views/9e9a5fa54bda03f0e68882d8c6206ec9.php
1:44:53 PM [vite] page reload storage/framework/views/9e9a5fa54bda03f0e68882d8c6206ec9.php (x2)
1:44:53 PM [vite] page reload storage/framework/views/f66072e9cb84663bc36ce8fef15beff1.php
1:44:53 PM [vite] page reload storage/framework/views/f66072e9cb84663bc36ce8fef15beff1.php (x2)
1:44:53 PM [vite] page reload storage/framework/views/9736886e7cc9e76213da34abbb824c5d.blade.php
1:44:53 PM [vite] page reload storage/framework/views/9736886e7cc9e76213da34abbb824c5d.blade.php (x2)
1:44:53 PM [vite] page reload storage/framework/views/05ca59f9ad738d4848d489b13b150f27.php
1:44:53 PM [vite] page reload storage/framework/views/e26246f9466c1d14fbfd6d9b13013415.php
1:44:53 PM [vite] page reload storage/framework/views/e26246f9466c1d14fbfd6d9b13013415.php (x2)
1:44:53 PM [vite] page reload storage/framework/views/05ca59f9ad738d4848d489b13b150f27.php
1:44:53 PM [vite] page reload storage/framework/views/e1a966cc57461255092d77780fef597d.php
1:44:53 PM [vite] page reload storage/framework/views/e1a966cc57461255092d77780fef597d.php (x2)
1:44:53 PM [vite] page reload storage/framework/views/3b86036c62a65cf58dadb3517288a587.php
1:44:53 PM [vite] page reload storage/framework/views/3b86036c62a65cf58dadb3517288a587.php (x2)
1:44:56 PM [vite] page reload storage/framework/views/05ca59f9ad738d4848d489b13b150f27.php
1:45:16 PM [vite] page reload storage/framework/views/f31e7e29187bdf109c52f19995dc9b97.php
1:45:16 PM [vite] page reload storage/framework/views/b8377a1732256871ab1f1b301cc73c56.blade.php
1:45:16 PM [vite] page reload storage/framework/views/39fb39a7cfe3e2f6c3fbf7f44a624713.php
1:45:16 PM [vite] page reload storage/framework/views/d3c6334bdf3afdd5a0a6e66f2ad22df6.php
1:45:16 PM [vite] page reload storage/framework/views/39fb39a7cfe3e2f6c3fbf7f44a624713.php
1:45:16 PM [vite] page reload storage/framework/views/d3c6334bdf3afdd5a0a6e66f2ad22df6.php
1:45:16 PM [vite] page reload storage/framework/views/c7a6bd8462ebefe8545c32d7f3223300.php
1:45:16 PM [vite] page reload storage/framework/views/c7a6bd8462ebefe8545c32d7f3223300.php (x2)
1:45:16 PM [vite] page reload storage/framework/views/b8377a1732256871ab1f1b301cc73c56.blade.php
1:45:16 PM [vite] page reload storage/framework/views/f31e7e29187bdf109c52f19995dc9b97.php
1:45:34 PM [vite] page reload storage/framework/views/39fb39a7cfe3e2f6c3fbf7f44a624713.php
1:45:43 PM [vite] page reload storage/framework/views/d173b7fc142a65c74cc3dfd63d20c11f.blade.php
1:45:43 PM [vite] page reload storage/framework/views/d173b7fc142a65c74cc3dfd63d20c11f.blade.php (x2)
1:45:43 PM [vite] page reload storage/framework/views/c271ab4894266a7ad37d442713fb08f0.php
1:45:43 PM [vite] page reload storage/framework/views/c271ab4894266a7ad37d442713fb08f0.php (x2)
1:45:43 PM [vite] page reload storage/framework/views/c97da0b552540c33409a64cffdc7e450.php
1:45:43 PM [vite] page reload storage/framework/views/c97da0b552540c33409a64cffdc7e450.php (x2)
1:46:03 PM [vite] page reload storage/framework/views/a5965e9990bc1ba1ed08a12038db9c34.blade.php
1:46:03 PM [vite] page reload storage/framework/views/a5965e9990bc1ba1ed08a12038db9c34.blade.php (x2)
1:46:03 PM [vite] page reload storage/framework/views/3d549831d80df3f8204f1671a329ac5b.php
1:46:03 PM [vite] page reload storage/framework/views/3d549831d80df3f8204f1671a329ac5b.php (x2)`

**Reply by ju5t** [reply]
@waapatricio it will only hot swap resources/assets that are defined like `@vite('resources/css/app.css')`. Everything else will cause a full page reload. This is expected behaviour and this happens on every page in your app. 

This cannot be changed, unless Vite comes up with a Livewire-way of reloading parts of a page.

To emphasise: this is not an issue with this package, it's a Vite-thing.
